pkg/apis/azure/validation/infrastructure.go

```diff
@@ -29,6 +31,44 @@ const (
 	natGatewayMaxTimeoutInMinutes int32 = 120
 )
 
+func ValidateInfrastructureConfigAgainstCloudProfile(oldInfra, infra *apisazure.InfrastructureConfig, shootRegion string, cloudProfile *gardencorev1beta1.CloudProfile, fld *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(infra.Networks.Zones) == 0 {
+		return allErrs
+	}
+
+	for _, region := range cloudProfile.Spec.Regions {
+		if region.Name == shootRegion {
+			allErrs = append(allErrs, validateInfrastructureConfigZones(oldInfra, infra, region.Zones, fld.Child("networks").Child("zones"))...)
+			break
+		}
+	}
+
+	return allErrs
+}
+
+func validateInfrastructureConfigZones(oldInfra, infra *apisazure.InfrastructureConfig, zones []gardencorev1beta1.AvailabilityZone, fld *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	azureZones := sets.NewString()
+	for _, zone := range zones {
+		azureZones.Insert(zone.Name)
+	}
+
+	for i, zone := range infra.Networks.Zones {
+		if oldInfra != nil && len(oldInfra.Networks.Zones) > i && oldInfra.Networks.Zones[i].Name == zone.Name {
```

Comment:
Yes, this is to prevent the case where removing the zone from the `CloudProfile` would render the shoot unable to update. If the zone was already part of the previous shoot configuration, we continue without checking if it still exists in the CP.